tao_compiler/mlir/disc/transforms/fusion_utils.cc

```diff
@@ -2521,5 +2828,828 @@ bool StitchCPUAnalysis::doCodeGeneration(OpBuilder& b, lmhlo::FusionOp fusion) {
+        }
+      }
+    }
+  }
+
+  if (fusion_type == FusionType::kNone || dominant_op == nullptr) {
+    return false;
+  } else if (fusion_type == FusionType::kStitch && !col_reductions.empty()) {
+    // Stitch fusion does not support col-reduction now.
+    return false;
+  }
+
+  fusion_pattern.setDominantOp(dominant_op);
+  fusion_pattern.setFusionType(fusion_type);
+  if (fusion_type == FusionType::kStitch) {
+    fusion_pattern.setSubRootOps(row_reductions);
+  }
+
+  return true;
+}
+
+bool StitchGpuFusionStrategy::tileXroots(ShapeAnalysis& shapeAnalysis,
+                                         FusionPattern& fusion_pattern) {
+  DenseMap<Value, TileInfo>& tile_plan = fusion_pattern.getTilePlan();
+  tile_plan.clear();
+
+  // Check constraint: all sub-roots have equivalent tiled dims and equivalent
+  // non-tiled dims. As all sub-roots are rank-2 row-reductions, the checking is
+  // simplified to check shape equality.
+  auto subroots = fusion_pattern.getSubRootOps();
+  auto dominant_op = fusion_pattern.getDominantOp();
+  if (!llvm::all_of(subroots, [&](Operation* op) {
+        if (op == dominant_op) {
+          return true;
+        } else {
+          return shapeAnalysis.isShapeEqual(dominant_op->getOperand(0),
+                                            op->getOperand(0));
+        }
+      })) {
+    LLVM_DEBUG(llvm::dbgs() << "Sub-roots do not have the same shape.");
+    return false;
+  }
+
+  // Tile subroots (row reductions).
+  for (Operation* op : subroots) {
+    // Tile input dimentions for input.
+    auto& in = tile_plan[op->getOperand(0)];
+    auto reduce = cast<lmhlo::ReduceOp>(op);
+    auto dimensions = reduce.dimensions().getValues<int64_t>();
+    for (auto& en : llvm::enumerate(dimensions)) {
+      in.tileSizes[en.value()] = ShapedType::kDynamicSize;
+    }
+    // No tile dimention for output.
+    auto& out = tile_plan[op->getOperand(2)];
+  }
+
+  // Tile non-subroot results. Successfully tiled roots are identified as
+  // regular xroots, while other roots are identified as irregular xroots.
+  auto& regular_xroots = fusion_pattern.getRegularXroots();
+  auto& irregular_xroots = fusion_pattern.getIrregularXroots();
+  regular_xroots.insert(subroots.begin(), subroots.end());
+  DenseSet<Operation*> subroots_set(subroots.begin(), subroots.end());
+  auto dominant = fusion_pattern.getDominantOp()->getOperand(0);
+  auto& dominant_tile = tile_plan[dominant];
+  auto results = fusion_pattern.getResults();
+  for (auto res : results) {
+    bool irregular = false;
+    Operation* op = fusion_pattern.findLastWriter(res);
+    if (subroots_set.contains(op)) {
+      continue;
+    }
+    // Find equal dims between res and a sub-root.
+    SmallVector<std::pair<SmallVector<int64_t>, SmallVector<int64_t>>> equal;
+    shapeAnalysis.extractContDimEqualInfo(dominant, res, equal);
+    int64_t dominant_non_tiled_dim_matched_n = 0;
+    DenseSet<int64_t> res_non_tiled_dims;
+    for (auto eq_dim : equal) {
```

Comment:
What if we got a squared row reduction op? that is ```[a, a] -> [a] ```. one not-tiled dim may be mapped to many different dims of a ranked value.

Reply:
Leave a TODO.